### PR TITLE
Blog post on using UDFs in python

### PR DESCRIPTION
**Pull request by timsaucer**
This PR adds a blog post describing using UDFs, and in particular on how to combine third party rust UDFs with datafusion-python.

### CODE REVIEW

File: _posts/2024-08-06-datafusion-python-udf-comparisons.md
Comment (kylebarron) [inline _posts/2024-08-06-datafusion-python-udf-comparisons.md:38]: You mention Polars but then in the next line come back to DataFusion. It might be worth a line to say why you decided on DataFusion over Polars, since they seem to have relatively similar goals here. Or maybe take out the polars reference if this blog post isn't really meant as a comparison to polars?

File: _posts/2024-08-06-datafusion-python-udf-comparisons.md
Comment (kylebarron) [inline _posts/2024-08-06-datafusion-python-udf-comparisons.md:48]: Nit: there's usually a hyphen for user-defined

File: _posts/2024-08-06-datafusion-python-udf-comparisons.md
Comment (kylebarron) [inline _posts/2024-08-06-datafusion-python-udf-comparisons.md:49]: Nit: should there be consistent casing for `DataFrame`?

File: _posts/2024-08-06-datafusion-python-udf-comparisons.md
Comment (kylebarron) [inline _posts/2024-08-06-datafusion-python-udf-comparisons.md:54]:
```suggestion
However, there will always arise times when you want a custom function. With user defined
```

File: _posts/2024-08-06-datafusion-python-udf-comparisons.md
Comment (kylebarron) [inline _posts/2024-08-06-datafusion-python-udf-comparisons.md:55]:
```suggestion
functions (UDFs) you open a world of possibilities in your code. Sometimes there simply isn’t an
```

File: _posts/2024-08-06-datafusion-python-udf-comparisons.md
Comment (kylebarron) [inline _posts/2024-08-06-datafusion-python-udf-comparisons.md:56]:
```suggestion
easy way to use built-in functions to achieve your goals.
```

File: _posts/2024-08-06-datafusion-python-udf-comparisons.md
Comment (kylebarron) [inline _posts/2024-08-06-datafusion-python-udf-comparisons.md:132]:
```suggestion
# Wrap our custom function with `datafusion.udf`, annotating expected 
# parameter and return types
is_of_interest = udf(
```

As a separate note, it wouldn't be hard to convert this `udf` function wrapper into a Python decorator, so we could do

```py
@udf(args=(pa.int64(), pa.int64(), pa.utf8()), returns=pa.bool_(), "stable")
def is_of_interest(
    partkey_arr: pa.Array,
    suppkey_arr: pa.Array,
    returnflag_arr: pa.Array,
) -> pa.Array: ...
```

File: _posts/2024-08-06-datafusion-python-udf-comparisons.md
Comment (kylebarron) [inline _posts/2024-08-06-datafusion-python-udf-comparisons.md:166]:
```suggestion
DataFusion uses [Apache Arrow](https://arrow.apache.org/) as its in-memory data format. This can
```

File: _posts/2024-08-06-datafusion-python-udf-comparisons.md
Comment (kylebarron) [inline _posts/2024-08-06-datafusion-python-udf-comparisons.md:168]:
```suggestion
that [pyarrow](https://arrow.apache.org/docs/python/), the canonical Python Arrow implementation, provides a variety of
```

File: _posts/2024-08-06-datafusion-python-udf-comparisons.md
Comment (kylebarron) [inline _posts/2024-08-06-datafusion-python-udf-comparisons.md:183]: I think you can use `pyarrow.is_in` to speed this up, instead of doing an equality check multiple times: https://arrow.apache.org/docs/python/generated/pyarrow.compute.is_in.html

File: _posts/2024-08-06-datafusion-python-udf-comparisons.md
Comment (kylebarron) [inline _posts/2024-08-06-datafusion-python-udf-comparisons.md:231]:
```suggestion
Python. I know of two use cases where I would recommend this approach. The first is the case when
```

File: _posts/2024-08-06-datafusion-python-udf-comparisons.md
Comment (kylebarron) [inline _posts/2024-08-06-datafusion-python-udf-comparisons.md:235]:
```suggestion
It is possible that you can implement your function in Rust to give a speed improvement and then
```

File: _posts/2024-08-06-datafusion-python-udf-comparisons.md
Comment (kylebarron) [inline _posts/2024-08-06-datafusion-python-udf-comparisons.md:240]: Capitalization nits: up to you if you wish to update them
```suggestion
When deciding to use this approach, it’s worth considering how much you think you’ll actually
benefit from the Rust implementation to decide if it’s worth the additional effort to maintain and
deploy the Python wheels you generate. It is certainly not necessary for every use case.
```

File: _posts/2024-08-06-datafusion-python-udf-comparisons.md
Comment (kylebarron) [inline _posts/2024-08-06-datafusion-python-udf-comparisons.md:120]: I think it might be helpful here to describe your problem a little bit. Say what `partkey_arr` is representing, and how it relates to your `values_of_interest` above.

File: _posts/2024-08-06-datafusion-python-udf-comparisons.md
Comment (kylebarron) [inline _posts/2024-08-06-datafusion-python-udf-comparisons.md:246]:
```suggestion
You’ll need [maturin](https://github.com/PyO3/maturin) to build the project, and you must use release mode when building to get the expected performance.
```

File: _posts/2024-08-06-datafusion-python-udf-comparisons.md
Comment (kylebarron) [inline _posts/2024-08-06-datafusion-python-udf-comparisons.md:296]: Ah one big difference here is that you're hard-coding the values of interest in Rust at compile time rather than the user being able to supply it at runtime. There's probably in theory some ways around this, where you can export to rust a wrapper function that you would call with your `values_of_interest` from Python, but it could be a little tricky.

File: _posts/2024-08-06-datafusion-python-udf-comparisons.md
Comment (kylebarron) [inline _posts/2024-08-06-datafusion-python-udf-comparisons.md:348]: Oh, oops haha you caught me!

File: _posts/2024-08-06-datafusion-python-udf-comparisons.md
Comment (kylebarron) [inline _posts/2024-08-06-datafusion-python-udf-comparisons.md:496]:
```suggestion
but that memory allocation is unnecessary. After that we use two `zip` operations so that we can
```

File: _posts/2024-08-06-datafusion-python-udf-comparisons.md
Comment (kylebarron) [inline _posts/2024-08-06-datafusion-python-udf-comparisons.md:485]: This is faster I suppose because it's not doing a boolean check on each individual array in its entirety and _then_ ORing them? It's doing it all at once in a single pass?

File: _posts/2024-08-06-datafusion-python-udf-comparisons.md
Comment (kylebarron) [inline _posts/2024-08-06-datafusion-python-udf-comparisons.md:594]:
```suggestion
[online documentation](https://datafusion.apache.org/python/) and jump in to help with one of the
```

File: _posts/2024-08-06-datafusion-python-udf-comparisons.md
Comment (timsaucer) [reply _posts/2024-08-06-datafusion-python-udf-comparisons.md:142]: Great idea. I've added it to the issue list https://github.com/apache/datafusion-python/issues/806

File: _posts/2024-08-06-datafusion-python-udf-comparisons.md
Comment (timsaucer) [reply _posts/2024-08-06-datafusion-python-udf-comparisons.md:496]: Yes, I didn't dive any deeper but my expectation is that by doing a single pass through the iteration we'll get a small speed improvement. It my modest test it only accounted for about a 5% boost.

File: _posts/2024-08-06-datafusion-python-udf-comparisons.md
Comment (timsaucer) [reply _posts/2024-08-06-datafusion-python-udf-comparisons.md:134]: I had a smaller statement in the earlier section, but I've expanded it because it was easy to pass over.

**Comment by timsaucer** [comment]
Huge tip of the hat to @kylebarron for the thorough feedback!

File: _posts/2024-08-06-datafusion-python-udf-comparisons.md
Comment (emgeee) [inline _posts/2024-08-06-datafusion-python-udf-comparisons.md:521]: should be "will contain all the values..."

File: _posts/2024-08-06-datafusion-python-udf-comparisons.md
Comment (emgeee) [inline _posts/2024-08-06-datafusion-python-udf-comparisons.md:530]: small orders should be anything below 5000

**Comment by timsaucer** [comment]
I keep meaning to get back to it and add a portion on window functions but I get distracted with other things. Yes, let me brush it up.

**Comment by emgeee** [comment]
I definitely found the post helpful!

**Review by andygrove — APPROVED** [review]
Excellent blog post! Thanks @timsaucer

File: _posts/2024-11-19-datafusion-python-udf-comparisons.md
Comment (alamb) [inline _posts/2024-11-19-datafusion-python-udf-comparisons.md:101]: 100% agree

BTW this is the kind of project I think is needed to make such queries much faster: https://github.com/apache/datafusion/issues/7955

File: _posts/2024-11-19-datafusion-python-udf-comparisons.md
Comment (alamb) [inline _posts/2024-11-19-datafusion-python-udf-comparisons.md:95]: I don't think we should change this blog post / example, but I wanted to point out that you can do this same calculation in SQL

```
> create table foo (x int, y int) as values (1,2), (3,4), (5,6);
0 row(s) fetched.
Elapsed 0.046 seconds.

> select * from foo where (x,y) IN ((3,4), (4,5));
+---+---+
| x | y |
+---+---+
| 3 | 4 |
+---+---+
1 row(s) fetched.
Elapsed 0.018 seconds.

> select * from foo where (x,y) IN ((3,4), (4,5));
+---+---+
| x | y |
+---+---+
| 3 | 4 |
+---+---+
1 row(s) fetched.
Elapsed 0.018 seconds.
```

DataFusion rewrites it internally using `struct`s:

```sql
> select * from foo where {'x': x, 'y': y} IN ({'x': 3, 'y': 4}, {'x': 4, 'y': 5});
+---+---+
| x | y |
+---+---+
| 3 | 4 |
+---+---+
```

```sql
> explain select * from foo where (x,y) IN ((3,4), (4,5));
+---------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| plan_type     | plan                                                                                                                                                                                                                                                                                                                                                       |
+---------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| logical_plan  | Filter: CAST(struct(foo.x, foo.y) AS Struct([Field { name: "c0", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "c1", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }])) IN ([Struct({c0:3,c1:4}), Struct({c0:4,c1:5})])                                         |
|               |   TableScan: foo projection=[x, y]                                                                                                                                                                                                                                                                                                                         |
| physical_plan | CoalesceBatchesExec: target_batch_size=8192                                                                                                                                                                                                                                                                                                                |
|               |   FilterExec: CAST(struct(x@0, y@1) AS Struct([Field { name: "c0", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "c1", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }])) IN ([Literal { value: Struct({c0:3,c1:4}) }, Literal { value: Struct({c0:4,c1:5}) }]) |
|               |     MemoryExec: partitions=1, partition_sizes=[1]                                                                                                                                                                                                                                                                                                          |
|               |                                                                                                                                                                                                                                                                                                                                                            |
+---------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
2 row(s) fetched.
Elapsed 0.003 seconds.
```

File: _posts/2024-11-19-datafusion-python-udf-comparisons.md
Comment (alamb) [inline _posts/2024-11-19-datafusion-python-udf-comparisons.md:250]: these are really good points

File: _posts/2024-11-19-datafusion-python-udf-comparisons.md
Comment (comphead) [inline _posts/2024-11-19-datafusion-python-udf-comparisons.md:38]:
```suggestion
[Polars](https://pola.rs/). The latter is growing extremely fast, and it serves many of the same use cases
```

File: _posts/2024-11-19-datafusion-python-udf-comparisons.md
Comment (comphead) [inline _posts/2024-11-19-datafusion-python-udf-comparisons.md:616]: Some of users jump directly into conclusions sections, so I'd add an short notice, what was speed benefits

File: _posts/2024-11-19-datafusion-python-udf-comparisons.md
Comment (alamb) [inline _posts/2024-11-19-datafusion-python-udf-comparisons.md:572]: A classic example is "avg" where the state is a sum + count and the output is a value

**Review by comphead — APPROVED** [review]
lgtm thanks @timsaucer

File: _posts/2024-11-19-datafusion-python-udf-comparisons.md
Comment (alamb) [inline _posts/2024-11-19-datafusion-python-udf-comparisons.md:603]: I think technically using pyarrow doesn't use rust -- it is implemented in C/C++. Possibly this would be slightly more accurate:

```suggestion
into using any functions that keep the data entirely on the Native (Rust or C/C++) side we see a near 10x speed
```

**Review by alamb — APPROVED** [review]
Thank you @timsaucer  -- I think this is great (and I certainly learned a lot) 

Once you merge this PR we need to do the publish process here: https://github.com/apache/datafusion-site?tab=readme-ov-file#publish-site

I am happy to do so as well, but I wanted to give you a heads up that it wouldn't just auto publish (no one has hooked that up yet)

File: _posts/2024-11-19-datafusion-python-udf-comparisons.md
Comment (timsaucer) [reply _posts/2024-11-19-datafusion-python-udf-comparisons.md:572]: `avg` was my first example before I found a bug (https://github.com/apache/datafusion-python/issues/797)  and switched it to sum to unblock myself!

**Comment by timsaucer** [comment]
Thank you all for the feedback. I've incorporated the suggestions and will wait to the end of my work day to see if there are any more comments before publishing.

**Comment by timsaucer** [comment]
I wasn't able to get the site to build on my machine. I'm going to try setting up a docker container just for building it tomorrow.

**Pull request merged**
